v4l2multi_stream_mmal low res mjpeg to frontend

popoviciri opened the pull request:
This PR simply brings the mjpeg options in the frontend, when using FNC `v4l2mmal_multi_stream` app written by @jasaw. These are `mjpegbitrate`, `mjpegwidth`, `mjpegheight` and `mjpegframerate` complementing #2120. A screenshot:
![image](https://user-images.githubusercontent.com/5729304/66274585-99bfab80-e86f-11e9-8257-8e6f5f945263.png)
Recommended merging order
1. #2117 
2. #2126 
3. This PR (there will be conflicts to sort out, some additional changes to `streameye.sh` I couldn't resist).

Looking forward to have this implemented upstream.
Cheers!

popoviciri committed:
add mjpeg options config readouts to streameye.sh

popoviciri committed:
add mjpeg options config readouts to frontend

jasaw commented:
@popoviciri I have just pushed a change to v4l2multi_stream_mmal mjpeg bitrate option. Previously the specified mjpeg bitrate was used directly in the mjpeg encoder which is very different from the mjpeg output stream bitrate. Now the mjpeg bitrate actually sets the output bitrate, and the internal mjpeg encoder bitrate is calculated from mjpeg_bitrate*video_framerate/mjpeg_framerate, and it's clipped at a maximum of 25Mbps.

Unfortunately, this affects your PR, which sets the default mjpeg bitrate at 2000000. Sorry about that.

jasaw commented on board/raspberrypi/motioneye-modules/streameyectl.py:
```diff
@@ -412,9 +412,13 @@ def _get_streameye_settings(camera_id):
+        'seMJPEGWidth': 640,
+        'seMJPEGHeight': 480,
+        'seMJPEGFramerate': 5,
+        'seMJPEGBitrate': 1000000
```
New default mjpeg bitrate of 800000. Sorry.

jasaw commented on board/raspberrypi/motioneye-modules/streameyectl.py:
```diff
@@ -442,15 +446,40 @@ def _get_streameye_settings(camera_id):
+    s.setdefault('seMJPEGWidth', 640)
+    s.setdefault('seMJPEGHeight', 480)
+    s.setdefault('seMJPEGFramerate', 5)
+    s.setdefault('seMJPEGBitrate', 2000000)
```
New default mjpeg bitrate of 800000. Sorry.

jasaw commented on board/raspberrypi/motioneye-modules/streameyectl.py:
```diff
@@ -1193,6 +1226,70 @@ def seRTSPPort():
+
+    return {
+        'label': 'MJPEG Bitrate',
+        'description': 'the MJPEG Bitrate fed to the frontend and used for motion detection on remote machines when streaming RTSP; for motion detection, 2000000 works good',
```
New default mjpeg bitrate of 800000. Sorry.

jasaw commented on board/raspberrypi/motioneye-modules/streameyectl.py:
```diff
@@ -1193,6 +1226,70 @@ def seRTSPPort():
+        'label': 'MJPEG Bitrate',
+        'description': 'the MJPEG Bitrate fed to the frontend and used for motion detection on remote machines when streaming RTSP; for motion detection, 2000000 works good',
+        'type': 'number',
+        'min': 1000000,
```
This will have to be reduced.

popoviciri committed:
consolidate mjpeg defaults

popoviciri committed:
parametric --intra input in v4l2mmal_multi_stream_app;
GOP = 2 * framerate

popoviciri commented:
Thanks @jasaw. I changed the defaults and consolidated the values across the two files.
I also pushed a commit where the --intra parameter is calculated from the framerate (double) as you suggested. This needs `bc` which although it is not in the default config, it is being installed by one of the apps so it works.
If you think is too risky to break later (say `bc` will not always be installed) let me know and I revert this to let --intra 50 hardcoded.
Cheers!

jasaw commented on board/raspberrypi/overlay/usr/bin/streameye.sh:
```diff
@@ -143,70 +43,100 @@ function invalid_opt() {
-        iptables -D INPUT -p tcp -s localhost --dport ${RTSP_PORT} -j ACCEPT
+            mjpeg_opts="--mjpegbitrate ${MJPEG_BITRATE} --mjpegwidth ${MJPEG_WIDTH} --mjpegheight ${MJPEG_HEIGHT} --mjpegframerate ${MJPEG_FRAMERATE}"
+            let "GOP=$((video_framerate))*2"
+            raspimjpeg_opts="${raspimjpeg_opts} --intra $GOP ${mjpeg_opts}"
```
We should probably check whether `intra` is specified in `raspimjpeg.conf` file before setting a default of `framerate*2` because line 94 will pick up the `intra` config.

jasaw commented on board/raspberrypi/motioneye-modules/streameyectl.py:
```diff
@@ -59,7 +59,7 @@
     ('flash', 'Flash'),
     ('horizon', 'Horizon'),
-    ('greyworld', 'Greyworld')
+    ('greyworld', 'Greyworld')                        
```
Just a nitpick... I would remove the trailing spaces.

jasaw commented:
This PR looks good. I like the idea of defaulting the `intra` to `2*framerate`, just need to check against `raspimjpeg.conf` file as mentioned.

Thanks for your hard work !

jasaw commented:
@popoviciri I have reverted `streameye.sh` changes from my PR in favour of yours to avoid unnecessary merge conflict.

popoviciri committed:
cleanup trailing spaces from streameyectl

popoviciri committed:
added checks:
- check whether intra is specified in raspimjpeg.conf; if NOT, then intra = 2 * framerate
- check is v4l2loopback module is loaded before trying to remove it

popoviciri commented:
Hi @jasaw. This is in fact your work. I merely made minor changes to what you already pushed here. Anyway, I made the required changes and added a check for v4l2loopback module before calling rmmod. Thanks again!

jasaw commented:
PR looks good. @ccrisan Have you reviewed this PR?

ccrisan commented:
Looks good to me. Let's have it merged :)
